### ISSUE TITLE
False-positive getMockBuilder abstract classes

### ISSUE THREAD

**Issue by MaximilianKresse** [kalessil/phpinspectionsea#1057]
| Subject        | Details                                                                       |
| :------------- | :---------------------------------------------------------------------------- |
| Issue type     | False-Positive                          |
| Plugin         | Php Inspections (EA Extended), 3.0.6.1 |
| Language level | PHP 7.1                        |

Current behaviour 
--
In PhpUnit Testcases I get a "Probable bugs" *error* when using TestCase::getMockBuilder() with abstract classes.

Expected behaviour
--
There are 3 methods:
- getMock
- getAbstractMock
- getMockBuilder

The difference between the first two methods and the last is the first two have many parameters and the later splits it into a chainable object structure and is much more readable. The first method is actually already removed for quite some time (no idea why the second still exist).

The last call on the MockBuilder actually decides what kind of mock you need "getMock", "getMockForAbstractClass" or "getMockForTrait". So the error to use getMockForAbstractClass simply wrong and should be quite the oposite, so you should be warned if you don't use the mock builder.

**Comment by kalessil** [comment]
Thank you for reporting an issue @Bob4ever, unfortunately I'm not following the description. Would you please provide a sample code or a screenshot which I could map with the description?

**Comment by Bob4ever** [comment]
Yeah sure:

```
<?php
declare(strict_types=1);

require_once __DIR__ . '/../vendor/autoload.php';

abstract class AbstractTest
{}

class Issue1057 extends \PHPUnit\Framework\TestCase
{
    /**
     * @throws ReflectionException
     */
    public function test()
    {
        // throws the error "Perhaps it was intended to mock it with getMockForAbstractClass method."
        $this->getMockBuilder(AbstractTest::class)->getMockForAbstractClass();
        // throws no error or warning
        $this->getMockForAbstractClass(AbstractTest::class);

        // why getMockForAbstractClass sucks

        $this->getMockForAbstractClass(AbstractTest::class, ['a'], '', false, false, false, [], true);
        // vs.
        $this->getMockBuilder(AbstractTest::class)
            ->setMethods(['a'])
            ->disableOriginalConstructor()
            ->disableOriginalClone()
            ->disableAutoload()
            ->enableArgumentCloning()
            ->getMockForAbstractClass();
    }
}
```

The bottom part is just an addition to represent why enforcing getMockForAbstractClass is not always a good idea. It is really hard to recognize what each flag does. (without the show argument name feature of phpstorm)

**Comment by Bob4ever** [comment]
Any news here? Do you understand my problem with the current behaviour?

**Comment by kalessil** [comment]
Hi @Bob4ever: yes, thank you for the example. Considering the issues to be a false-positive - seems we can easily recognize the context from the example.

**Comment by leofeyer** [comment]
Here are some additional screenshots:

<img width="647" alt="" src="https://user-images.githubusercontent.com/1192057/50642321-0cab9700-0f6b-11e9-8f7a-9564f8d3d90b.png">

<img width="617" alt="" src="https://user-images.githubusercontent.com/1192057/50642336-1a611c80-0f6b-11e9-9bd1-3081bae4b2c6.png">

**Comment by kalessil** [comment]
Fixed!
